### ISSUE TITLE
misleading error message when type with index signature assigned to "compatible" non-signature type

### ISSUE THREAD

**Issue by tvald** [Microsoft/TypeScript#28193]
A misleading error message (`property 'bar' is missing...`) is produced when attempting to assign a type with an index signature to a seemingly-compatible type without an index signature. After much searching, I think the incompatibility is by design, so the error message should clearly indicate that design choice.

**TypeScript Version:**  3.1.3

**Search Terms:**
property is missing in type record index signature indexed mapped explicit property field assign convert generic specific

**Code**

```ts
const record: Record<string, string> = { bar: '' };
const indexed: { [key: string]: string } = { bar: '' };

let foo: { bar: string };
foo = record;  // error: property 'bar' is missing...
foo = indexed; // error: property 'bar' is missing...
```

**Expected behavior:**
An error message indicating that a type with an index signature is not assignable to a type without an index signature. eg, (*Edit: added both error messages*)
```
Type 'Record<string, string>' is not assignable to type '{ bar: string; }'.
  Index signature is missing in type '{ bar: string; }'.
```
```
Type '{ [key: string]: string; }' is not assignable to type '{ bar: string; }'.
  Index signature is missing in type '{ bar: string; }'.
```

**Actual behavior:**
*Edit: added both error messages*
```
Type 'Record<string, string>' is not assignable to type '{ bar: string; }'.
  Property 'bar' is missing in type 'Record<string, string>'.
```
```
Type '{ [key: string]: string; }' is not assignable to type '{ bar: string; }'.
  Property 'bar' is missing in type '{ [key: string]: string; }'.
```

**Workaround**
Explicit conversion works, if you can name or define the type you're assigning to:
```ts
foo = record as { bar: string };
foo = indexed as { bar: string };
```

**Playground Link:**
[playground link](https://www.typescriptlang.org/play/#src=const%20record%3A%20Record%3Cstring%2C%20string%3E%20%3D%20%7B%20bar%3A%20''%20%7D%3B%0D%0Aconst%20indexed%3A%20%7B%20%5Bkey%3A%20string%5D%3A%20string%20%7D%20%3D%20%7B%20bar%3A%20''%20%7D%3B%0D%0A%0D%0Alet%20foo%3A%20%7B%20bar%3A%20string%20%7D%3B%0D%0Afoo%20%3D%20record%3B%0D%0Afoo%20%3D%20indexed%3B%0D%0A)

**Related Issues:** 
#12662, #17223, #21121

> [mhegazy commented in #21121](https://github.com/Microsoft/TypeScript/issues/21121#issuecomment-356692449)
> Types are open in TS. So SpecificFields could be assigned a value that say has other: string.
> Type GenericFields has a stronger constraint, it says all properties has to be of type boolean.

> [RyanCavanaugh commented in #12662](https://github.com/Microsoft/TypeScript/issues/12662#issuecomment-264997163)
> The index signature is exclusionary here because it implies the non-existence of other explicit properties.

**Comment by a-tarasyuk** [comment]
@tvald @weswigham 

Should both versions produce the same error message?

```shell
Type 'Record<string, string>' is not assignable to type '{ bar: string; }'.
  Index signature is missing in type '{ bar: string; }'.


Type '{ [key: string]: string; }' is not assignable to type '{ bar: string; }'.
  Index signature is missing in type '{ bar: string; }'.

```

Or only `indexed`?

**Comment by tvald** [comment]
Yes, since `Record<string, ...>` is equivalent to an explicit index signature. I'm not sure how things work under the hood, though, so it might require a couple of different patches?

**Comment by a-tarasyuk** [comment]
@weswigham based on the code, these changes should be add here, correct me if I'm wrong

```typescript
// checker.ts

const unmatchedProperty = getUnmatchedProperty(source, target, requireOptionalProperties);
if (unmatchedProperty) {
    if (reportErrors) {
        reportError(Diagnostics.Property_0_is_missing_in_type_1, symbolToString(unmatchedProperty), typeToString(source));
    }
    return Ternary.False;
}
```
however, if we change the message, it will affect other parts which handled by this logic, for instance

```typescript
class C implements Function {
    prototype = null;
    length = 1;
    arguments = null;
    caller = () => { };
}
```

Before 

```shell
Class 'C' incorrectly implements interface 'Function'.
  Property 'apply' is missing in type 'C'.
```

After
```shell
Class 'C' incorrectly implements interface 'Function'.
  Index signature is missing in type 'Function'.
```

I think the previous message should not be changed. Is there any way to handle only two cases?

**Comment by weswigham** [comment]
Inside the `if (reportErrors) {` branch you'll want to inspect the `source` and `target` types to determine which error message is appropriate.

**Comment by tvald** [comment]
Oh hmm. [playground link](https://www.typescriptlang.org/play/#src=declare%20const%20indexed%3A%20%7B%20%5Bkey%3A%20string%5D%3A%20string%3B%20bar%3A%20string%20%7D%3B%0D%0A%0D%0Alet%20foo%3A%20%7B%20bar%3A%20string%20%7D%3B%0D%0Afoo%20%3D%20indexed%3B%0D%0A) for the example by @DanielRosenwasser, which compiles without error.

I suppose the error message is accurate, in that the type with an index signature becomes assignable if the property is explicitly declared. However, that seems to contradict the reasoning (as I understood it) that the index signature is not assignable due to a "stronger constraint" and an "exclusionary" nature.

Eh... I don't see a clear change that would improve the status quo, so hopefully this issue will just document what's going on for anyone who runs into it in the future. My preference would have been to just allow assignment of the index signature type in the first place. /shrug

Feel free to reopen if anyone thinks there is a valuable change to be had.
